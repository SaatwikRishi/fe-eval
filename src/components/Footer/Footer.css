
footer {
    display: flex;
    padding: 30px;   
    min-height: 70px;
    width: 100%;
    display: flex;
    justify-content: space-between;

}
.footer-themes {
    display: flex;
    column-gap: 20px;
    color: white;
display: flex;
    background-color: black;
    align-items: center;

}

