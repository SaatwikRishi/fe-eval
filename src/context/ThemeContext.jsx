import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { makeRequest } from '../util/makeRequest'
import { GET_THEMES } from '../util/constants'
import { useNavigate } from 'react-router-dom'

export const ThemeContext = React.createContext()

export default function ThemeState ({ children }) {
  const [themes, setThemes] = React.useState([])
  const [loading, setLoading] = React.useState(true)
  const [currentTheme, setCurrentTheme] = React.useState(1)
  const getCurrentTheme = () => {
    const th = themes.filter((e) => e.id === currentTheme)[0].colorHexCode
    return th
  }

  const navigate = useNavigate()

  useEffect(() => {
    makeRequest(GET_THEMES())
      .then((e) => {
        setCurrentTheme(e.preferredThemeId)

        setThemes(e.themes)
        setTimeout(() => {
          setLoading(false)
        })
      }, 300)
      .catch((e) => {
        navigate('/error/500')
        setLoading(false)
      })
  }, [])
  if (loading) return <div>Loading</div>
  return (
    <ThemeContext.Provider value={{ themes, setCurrentTheme, getCurrentTheme }}>
      {children}
    </ThemeContext.Provider>
  )
}

ThemeState.propTypes = {
  children: PropTypes.node.isRequired
}
