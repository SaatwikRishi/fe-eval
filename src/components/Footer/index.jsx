import React, { useContext } from 'react'
import { ThemeContext } from '../../context/ThemeContext'
import { UPDATE_THEME } from '../../util/constants'
import { makeRequest } from '../../util/makeRequest'
import CustomButton from '../CustomButton'
import './Footer.css'
export default function Footer () {
  // const [themeClick, setThemeClick] = React.useState(null)
  const { getCurrentTheme, themes, setCurrentTheme } = useContext(ThemeContext)

  const [themeChanged, setThemeChanged] = React.useState(false)
  const handleThemeClick = (color) => {
    const nt = themes.filter((theme) => theme.colorHexCode === color)[0].id
    setCurrentTheme(nt)
    setThemeChanged(nt)
  }
  const handleThemeChange = () => {
    makeRequest(UPDATE_THEME(themeChanged)).then((e) => {
      setThemeChanged(false)
    })
  }

  return (
    <footer className='paddings'
      style={{
        backgroundColor: getCurrentTheme()
      }}
    >
      <div
        className="footer-themes"
        style={{
          backgroundColor: getCurrentTheme()
        }}
      >
        <h2>THEMES</h2>
        {themes.map(
          (color, index) =>
            color.colorHexCode !== getCurrentTheme() && (
              <div
                data-testid='change-theme'
                onClick={() => handleThemeClick(color.colorHexCode)}
                key={index}
                style={{
                  backgroundColor: color.colorHexCode,
                  width: '40px',
                  height: '40px',
                  borderRadius: '5px'
                }}
              ></div>
            )
        )}
      </div>
    {
themeChanged && <CustomButton text='SAVE THEME' onClick={handleThemeChange} />}
    </footer>
  )
}
