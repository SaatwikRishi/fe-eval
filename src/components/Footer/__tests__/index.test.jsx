import { render } from '@testing-library/react'
import Footer from '..'
import React from 'react'
import { ThemeContext } from '../../../context/ThemeContext'

describe('Tests for footer', () => {
  it('should render when component is called', () => {
    const { container } = render(
      <ThemeContext.Provider value={{
        getCurrentTheme: jest.fn(),
        themes: [
          {
            id: 1,
            colorHexCode: '#000000'
          },
          {
            id: 2,
            colorHexCode: '#800080'
          },
          {
            id: 3,
            colorHexCode: '#0000FF'
          },
          {
            id: 4,
            colorHexCode: '#9B9999'
          }
        ]
      }}>
        <Footer />
      </ThemeContext.Provider>
    )
    expect(container).toMatchSnapshot()
  })
  it('should show the save theme button when clicked', () => {
    const { getByTestId } = render(
      <ThemeContext.Provider value={{
        getCurrentTheme: jest.fn(),
        themes: [
          {
            id: 1,
            colorHexCode: '#000000'
          },
          {
            id: 2,
            colorHexCode: '#800080'
          },
          {
            id: 3,
            colorHexCode: '#0000FF'
          },
          {
            id: 4,
            colorHexCode: '#9B9999'
          }
        ]
      }}>
        <Footer />
      </ThemeContext.Provider>
    )
    const saveThemeButton = getByTestId('save-theme')
    expect(saveThemeButton).toBeTruthy()
  })
})
