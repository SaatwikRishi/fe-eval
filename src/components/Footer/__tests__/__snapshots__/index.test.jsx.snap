// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for footer should render when component is called 1`] = `
<div>
  <footer
    class="paddings"
  >
    <div
      class="footer-themes"
    >
      <h2>
        THEMES
      </h2>
      <div
        style="background-color: rgb(0, 0, 0); width: 40px; height: 40px; border-radius: 5px;"
      />
      <div
        style="background-color: rgb(128, 0, 128); width: 40px; height: 40px; border-radius: 5px;"
      />
      <div
        style="background-color: rgb(0, 0, 255); width: 40px; height: 40px; border-radius: 5px;"
      />
      <div
        style="background-color: rgb(155, 153, 153); width: 40px; height: 40px; border-radius: 5px;"
      />
    </div>
  </footer>
</div>
`;
